airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/Partitions.kt

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.read
+
+import java.util.concurrent.atomic.AtomicInteger
+import kotlinx.coroutines.sync.Mutex
+
+/** A [Mutex] wrapper that only allows locking n times */
+internal class LimitedLockMutex(maxCount: Int) {
+    private val mutex = Mutex()
+    private var locksLeft = AtomicInteger(maxCount)
+
+    @Synchronized
```

Comment:
`@Synchronized` annotation in kotlin generates a `synchronized` jvm function - it look like it maps directly.
Synchronized(this) { block } is similar to java's synchronized block I think.

Reply:
because locksLeft and mutex are thread safe they don't need protection themselves, I'm trying to protect from a situation one thread is check `canLock()` while another thread is in the middle of `tryLock()`.